File: hibernate-dialect/src/main/java/tech/ydb/hibernate/dialect/code/YdbJdbcCode.java
package tech.ydb.hibernate.dialect.code;

/**
 * @author Kirill Kurdyukov
 */
public final class YdbJdbcCode {

    /**
     * Boolean value.
     */
    public static final int BOOL = YdbConst.SQL_KIND_PRIMITIVE;

    /**
     * A signed integer. Acceptable values: from -2^7 to 2^7–1. Not supported for table columns
     */
    public static final int INT8 = YdbConst.SQL_KIND_PRIMITIVE + 1;

    /**
     * An unsigned integer. Acceptable values: from 0 to 2^8–1.
     */
    public static final int UINT8 = YdbConst.SQL_KIND_PRIMITIVE + 2;

    /**
     * A signed integer. Acceptable values: from –2^15 to 2^15–1. Not supported for table columns
     */
    public static final int INT16 = YdbConst.SQL_KIND_PRIMITIVE + 3;

    /**
     * An unsigned integer. Acceptable values: from 0 to 2^16–1. Not supported for table columns
     */
    public static final int UINT16 = YdbConst.SQL_KIND_PRIMITIVE + 4;

    /**
     * A signed integer. Acceptable values: from –2^31 to 2^31–1.
     */
    public static final int INT32 = YdbConst.SQL_KIND_PRIMITIVE + 5;

    /**
     * An unsigned integer. Acceptable values: from 0 to 2^32–1.
     */
    public static final int UINT32 = YdbConst.SQL_KIND_PRIMITIVE + 6;

    /**
     * A signed integer. Acceptable values: from –2^63 to 2^63–1.
     */
    public static final int INT64 = YdbConst.SQL_KIND_PRIMITIVE + 7;

    /**
     * An unsigned integer. Acceptable values: from 0 to 2^64–1.
     */
    public static final int UINT64 = YdbConst.SQL_KIND_PRIMITIVE + 8;

    /**
     * A real number with variable precision, 4 bytes in size. Can't be used in the primary key
     */
    public static final int FLOAT = YdbConst.SQL_KIND_PRIMITIVE + 9;

    /**
     * A real number with variable precision, 8 bytes in size. Can't be used in the primary key
     */
    public static final int DOUBLE = YdbConst.SQL_KIND_PRIMITIVE + 10;

    /**
     * A binary data, synonym for YDB type String
     */
    public static final int BYTES = YdbConst.SQL_KIND_PRIMITIVE + 11;

    /**
     * Text encoded in UTF-8, synonym for YDB type Utf8
     */
    public static final int TEXT = YdbConst.SQL_KIND_PRIMITIVE + 12;

    /**
     * YSON in a textual or binary representation. Doesn't support matching, can't be used in the primary key
     */
    public static final int YSON = YdbConst.SQL_KIND_PRIMITIVE + 13;

    /**
     * JSON represented as text. Doesn't support matching, can't be used in the primary key
     */
    public static final int JSON = YdbConst.SQL_KIND_PRIMITIVE + 14;

    /**
     * Universally unique identifier UUID. Not supported for table columns
     */
    public static final int UUID = YdbConst.SQL_KIND_PRIMITIVE + 15;

    /**
     * Date, precision to the day
     */
    public static final int DATE = YdbConst.SQL_KIND_PRIMITIVE + 16;

    /**
     * Date/time, precision to the second
     */
    public static final int DATETIME = YdbConst.SQL_KIND_PRIMITIVE + 17;

    /**
     * Date/time, precision to the microsecond
     */
    public static final int TIMESTAMP = YdbConst.SQL_KIND_PRIMITIVE + 18;

    /**
     * Time interval (signed), precision to microseconds
     */
    public static final int INTERVAL = YdbConst.SQL_KIND_PRIMITIVE + 19;

    /**
     * Date with time zone label, precision to the day
     */
    public static final int TZ_DATE = YdbConst.SQL_KIND_PRIMITIVE + 20;

    /**
     * Date/time with time zone label, precision to the second
     */
    public static final int TZ_DATETIME = YdbConst.SQL_KIND_PRIMITIVE + 21;

    /**
     * Date/time with time zone label, precision to the microsecond
     */
    public static final int TZ_TIMESTAMP = YdbConst.SQL_KIND_PRIMITIVE + 22;

    /**
     * JSON in an indexed binary representation. Doesn't support matching, can't be used in the primary key
     */
    public static final int JSON_DOCUMENT = YdbConst.SQL_KIND_PRIMITIVE + 23;

    public static final int DATE_32 = YdbConst.SQL_KIND_PRIMITIVE + 25;

    public static final int DATETIME_64 = YdbConst.SQL_KIND_PRIMITIVE + 26;

    public static final int TIMESTAMP_64 = YdbConst.SQL_KIND_PRIMITIVE + 27;

    public static final int INTERVAL_64 = YdbConst.SQL_KIND_PRIMITIVE + 28;


    // WARN! Attribute value must be constant!

    /**
     * <a href="https://github.com/ydb-platform/ydb-jdbc-driver/blob/v2.3.3/jdbc/src/main/java/tech/ydb/jdbc/impl/YdbTypes.java#L37-L66">link</a>
     */
    public static final int DECIMAL_22_9 = YdbConst.SQL_KIND_DECIMAL + (22 << 6) + 9;

    public static final int DECIMAL_31_9 = YdbConst.SQL_KIND_DECIMAL + (31 << 6) + 9;

    public static final int DECIMAL_35_0 = YdbConst.SQL_KIND_DECIMAL + (35 << 6);

    public static final int DECIMAL_35_9 = YdbConst.SQL_KIND_DECIMAL + (35 << 6) + 9;
}

File: kotlin-exposed-dialect/src/main/kotlin/tech/ydb/exposed/dialect/code/YdbJdbcCode.kt
package tech.ydb.exposed.dialect.code

/**
 * YDB JDBC vendor type codes for [java.sql.PreparedStatement.setObject] (index, value, sqlType).
 *
 * Layout matches the YDB JDBC driver ([YdbTypes](https://github.com/ydb-platform/ydb-jdbc-driver/blob/master/jdbc/src/main/java/tech/ydb/jdbc/common/YdbTypes.java)).
 */
internal object YdbJdbcCode {
    const val SQL_KIND_PRIMITIVE: Int = 10_000
    const val SQL_KIND_DECIMAL: Int = 1 shl 14

    const val BOOL: Int = SQL_KIND_PRIMITIVE
    const val INT8: Int = SQL_KIND_PRIMITIVE + 1
    const val UINT8: Int = SQL_KIND_PRIMITIVE + 2
    const val INT16: Int = SQL_KIND_PRIMITIVE + 3
    const val UINT16: Int = SQL_KIND_PRIMITIVE + 4
    const val INT32: Int = SQL_KIND_PRIMITIVE + 5
    const val UINT32: Int = SQL_KIND_PRIMITIVE + 6
    const val INT64: Int = SQL_KIND_PRIMITIVE + 7
    const val UINT64: Int = SQL_KIND_PRIMITIVE + 8
    const val FLOAT: Int = SQL_KIND_PRIMITIVE + 9
    const val DOUBLE: Int = SQL_KIND_PRIMITIVE + 10
    const val BYTES: Int = SQL_KIND_PRIMITIVE + 11
    const val TEXT: Int = SQL_KIND_PRIMITIVE + 12
    const val YSON: Int = SQL_KIND_PRIMITIVE + 13
    const val JSON: Int = SQL_KIND_PRIMITIVE + 14
    const val UUID: Int = SQL_KIND_PRIMITIVE + 15
    const val DATE: Int = SQL_KIND_PRIMITIVE + 16
    const val DATETIME: Int = SQL_KIND_PRIMITIVE + 17
    const val TIMESTAMP: Int = SQL_KIND_PRIMITIVE + 18
    const val INTERVAL: Int = SQL_KIND_PRIMITIVE + 19
    const val TZ_DATE: Int = SQL_KIND_PRIMITIVE + 20
    const val TZ_DATETIME: Int = SQL_KIND_PRIMITIVE + 21
    const val TZ_TIMESTAMP: Int = SQL_KIND_PRIMITIVE + 22
    const val JSON_DOCUMENT: Int = SQL_KIND_PRIMITIVE + 23
    const val DATE32: Int = SQL_KIND_PRIMITIVE + 25
    const val DATETIME64: Int = SQL_KIND_PRIMITIVE + 26
    const val TIMESTAMP64: Int = SQL_KIND_PRIMITIVE + 27
    const val INTERVAL64: Int = SQL_KIND_PRIMITIVE + 28

    fun decimal(precision: Int, scale: Int): Int {
        require(precision in 1..35) { "YDB Decimal precision must be in 1..35" }
        require(scale in 0..precision) { "YDB Decimal scale must be in 0..precision" }
        return SQL_KIND_DECIMAL + (precision shl 6) + scale
    }
}
